### DISCUSSION TITLE
Relocate css files to Azure Blob Storage

### DISCUSSION THREAD

**Discussion by danielfrom** [elementor/elementor#18987] [Developers Community]
Hi there,

Is it possible to change the path for the output css-files  that are generated by Elementor?
(Currently /wp-content/uploads/elementor/css/post-ID.css)

I'm using the "Microsoft Azure Storage for WordPress"-plugin for storing the uploads in Azure Blob Storage. It works with files from "any" plugins except from Elementor. The files are forced to be saved locally in the folder: /wp-content/uploads/elementor/css/ instead of the storage container set up with the plugin.

Thanks for your time,
Daniel

**Comment by ali-awwad** [comment]
I had the same issue. as a work around, i used w3total cache, there is an option to"host custom files"

Go to Performance > CDN , then under "Advanced" tab, find the "Custom file list", there add to the top two new lines
```
{wp_content_dir}/uploads/elementor/css/*
{plugins_dir}/elementor/assets/*
```

then on the same page, go up to "General" tab, and check "Host custom files" and click upload. after that you have to clear the caches.

Hope it helps.

**Comment by shane-js** [comment]
Similarly I am having related issue using a fresh Azure Wordpress using their Azure blobstorage CDN (I let it auto-configure W3 Total Cache - nothing custom on my end) and a popular theme built on Elementor.

I have done what @ali-awwad posted above but it does not work - no css files get created. Even if I attempt to regenerate them. In fact there is no "elementor" directory even made in the wp-content/uploads path in the blob storage account. The only thing I have found that works is if I switch 'CSS Print Method' from 'external file' to 'internal embedding'. 

When I am in the elementor editor for the page that 404s the CSS files I see it perfectly fine. Just the published version cannot find the files (and makes sense because I don't see them either as I mentioned).

It seems Azure Wordpress ships with "Organize my uploads into month- and year-based folders" so all my directories in uploads are year folders. Not sure if that is relevant.

**Reply by shane-js** [reply]
@nicholaszein the way you closed https://github.com/elementor/elementor/issues/25075 did not allow any replies/further/future comments. I am not sure if this was intentional. I had a hunch it would be closed as "not elementor" despite the info I added so I am not surprised on that front. However I am now thinking it may have nothing to do with Azure at all.  Furthermore it had nothing to do with directory permissions and didn't require the un-recommended (from the docs) css print method of internal embedding.

I am not a big wordpress person so don't have readily available dev flows to 100% prove this perhaps someone from Elementor will find it in the product's interest to scope this out or at least document it.  At the very least at least this information being here can help someone in the future save countless hours I wasted figuring this out.

Basically I have found now that I think it is a Wordpress media setting +  W3 with CDN (pretty common use case from my understanding) + Elementor conflict. To @ali-awwad point above you need to add the Elementor upload path to the custom files... but here is the catch.... if your Wordpress -> Settings -> Media -> Uploading Files -> "Organize my uploads into month- and year-based folders" is checked (on/enabled, which just happened to be the default for Azure Wordpress) then **no elementor post css file is ever created**.

I had a hunch because where elementor saves those uploads to was all year dated directories so I was thinking wordpress might be rejecting it because the path elementor saves those css files in doesn't follow that structure.  Once I turned off "Organize my uploads into month- and year-based folders" even **using Elementor's "Regenerate CSS & Data" did absolutely nothing**. Here is what did work though:

1. I had to manually go back into each page with a post css and change something css related and click "update" button.
2. From W3 run "Upload custom files" (if you ctrl+f "uploads" you should see the post css file you are looking for in the list that was uploaded after)
3. From W3 Empty All Caches

_Notes_: 

- Sometimes I had to do 1-3 multiple times - couldn't tell you why. 
- I had to turn on "Force over-writing of existing files" in W3 as well because I found sometimes even reloading custom files wouldn't pick up small changes like colors. 
- Also note that multiple times I found even hard refresh in incognito/private tab the version url param passed on loading the post css file would get stuck on an old one even if in the Azure blobstorage (in my case but whatever CDN you use) had the newest CSS changes - I found just closing chrome/brave and reopening was more foolproof.) 

At that point the CSS files finally were actually saved in the "wp-content/uploads/elementor/css/" directory for them to be read back correctly. At that point typical W3 cache flow comes in (such as need to reload custom files if you tweak css for a page for it to be updated in the CDN cache) but at least the file exists.  Any new pages made after disabling "Organize my uploads into month- and year-based folders" still required the above steps.
